### PR TITLE
gpu/ze: use stream load/store for GPU fast copy

### PR DESCRIPTION
**Pull request by zhenggb72**
## Pull Request Description

use stream load/store to speedup fast GPU copy

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [ ] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [ ] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by zhenggb72** [comment]
With stream load/store on PVC (osu-latency 2 GPUs, different sockets):

0                       1.59
1                       4.07
2                       3.65
4                       3.81
8                       3.64
16                      3.81
32                      3.63
64                      3.73
128                     3.63
256                     3.74
512                     4.27
1024                    4.42
2048                    5.27
4096                    6.83


before:

\# Send Buffer on DEVICE (D) and Receive Buffer on DEVICE (D)
\# Size          Latency (us)
0                       1.41
1                       3.91
2                       3.77
4                       3.91
8                       3.77
16                      3.91
32                      3.77
64                      3.84
128                     4.88
256                     7.32
512                    12.49
1024                   23.14
2048                    9.41
4096                    9.39

File: src/mpid/ch4/netmod/ofi/ofi_recv.h [hunk 1]
```diff
@@ -282,7 +282,8 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_do_irecv(void *buf,
 
         /* Unpack */
         MPIDI_OFI_REQUEST(rreq, event_id) = MPIDI_OFI_EVENT_RECV_PACK;
-        MPIDI_OFI_REQUEST(rreq, noncontig.pack.pack_buffer) = MPL_malloc(data_sz, MPL_MEM_OTHER);
+        MPIDI_OFI_REQUEST(rreq, noncontig.pack.pack_buffer) =
+            MPL_aligned_alloc(256, data_sz, MPL_MEM_OTHER);
```
Comment (yfguo) [inline src/mpid/ch4/netmod/ofi/ofi_recv.h:286]: I think the alignment parameter is byte. Are we sure we want to align to 256B?

File: src/mpid/ch4/netmod/ofi/ofi_send.h [hunk 2]
```diff
@@ -625,7 +625,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_send(const void *buf, MPI_Aint count, MPI
         /* inject path */
         void *pack_buf = NULL;
         if (need_pack) {
-            pack_buf = MPL_malloc(data_sz, MPL_MEM_OTHER);
+            pack_buf = MPL_aligned_alloc(256, data_sz, MPL_MEM_OTHER);
```
Comment (yfguo) [inline src/mpid/ch4/netmod/ofi/ofi_send.h:628]: Same question as above.

File: src/mpl/src/gpu/mpl_gpu_ze.c [hunk 3]
```diff
@@ -3414,7 +3445,7 @@ int MPL_gpu_fast_memcpy(void *src, MPL_pointer_attr_t * src_attr, void *dest,
     if (n == 1) {
         *(char *) d = *(char *) s;
     }
-#if defined(MPL_HAVE__MM256_STOREU_SI256) || defined(MPL_HAVE__MM512_STOREU_SI512)
+#if defined(MPL_HAVE__MM512_STOREU_SI512) || defined(MPL_HAVE__MM512_STREAM_SI512) || defined(MPL_HAVE__MM256_STREAM_SI256) || defined(MPL_HAVE__MM256_STOREU_SI256)
```
Comment (yfguo) [inline src/mpl/src/gpu/mpl_gpu_ze.c:3448]: I am wondering if the separate checks for `_mm512_stream_si512` and `_mm512_storeu_si512` are necessary. Is it possible for a compiler and architecture that support AVX512F (which we already checks) that does not have these two?

**Review by yfguo — CHANGES_REQUESTED** [review]

File: src/mpl/src/gpu/mpl_gpu_ze.c (hunk 3 above)
Comment (zhenggb72) [reply src/mpl/src/gpu/mpl_gpu_ze.c:3448]: I am not sure, autoconf tests both cases separately and define these two macro separately. So I think to be safe, they should be handled separately here as well.

File: src/mpid/ch4/netmod/ofi/ofi_recv.h (hunk 1 above)
Comment (zhenggb72) [reply src/mpid/ch4/netmod/ofi/ofi_recv.h:286]: Ok, I will change alignment to 64.

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch4/gpu

**Review by yfguo — APPROVED** [review]

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch4/gpu

**Comment by yfguo** [comment]
test:mpich/ch4/ofi
test:mpich/ch4/gpu

**Pull request merged**